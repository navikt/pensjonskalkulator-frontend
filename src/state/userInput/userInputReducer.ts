import { createSlice, PayloadAction } from '@reduxjs/toolkit'

import { formatInntekt } from '../../utils/inntekt'

export interface Simulation {
  formatertUttaksalderReadOnly: string | null // (!) Obs READONLY - string i format "YY alder.aar string.og M alder.maaneder" - oppdateres automatisk basert på uttaksalder - se uttaksalderListener
  uttaksalder: Alder | null // valgt uttaksalder for 100% alderspensjon (alder perioden gjelder FRA) - aar heltall, maaneder heltall mellom 0-11
  aarligInntektFoerUttakBeloep: string | null // inntekt før uttak av pensjon - formatert string i nok - overskriver beløp fra Skatteetaten
  aarligInntektVsaHelPensjon?: AarligInntektVsaPensjon
  gradertUttaksperiode: GradertUttak | null
}

export interface UserInputState {
  utenlandsopphold: boolean | null
  samtykke: boolean | null
  afp: AfpRadio | null
  samboer: boolean | null
  currentSimulation: Simulation
}

export const userInputInitialState: UserInputState = {
  utenlandsopphold: null,
  samtykke: null,
  afp: null,
  samboer: null,
  currentSimulation: {
    formatertUttaksalderReadOnly: null,
    uttaksalder: null,
    aarligInntektFoerUttakBeloep: null,
    gradertUttaksperiode: null,
  },
}

export const userInputSlice = createSlice({
  name: 'userInputSlice',
  initialState: userInputInitialState,
  reducers: {
    setUtenlandsopphold: (state, action: PayloadAction<boolean>) => {
      state.utenlandsopphold = action.payload
    },
    setSamtykke: (state, action: PayloadAction<boolean>) => {
      state.samtykke = action.payload
    },
    setAfp: (state, action: PayloadAction<AfpRadio>) => {
      state.afp = action.payload
    },
    setSamboer: (state, action: PayloadAction<boolean>) => {
      state.samboer = action.payload
    },
    setCurrentSimulationUttaksalder: (
      state,
      action: PayloadAction<{
        aar: number
        maaneder: number
      } | null>
    ) => {
      state.currentSimulation = {
        ...state.currentSimulation,
        uttaksalder: action.payload ? { ...action.payload } : null,
      }
    },
    setCurrentSimulationAarligInntektFoerUttakBeloep: (
      state,
      action: PayloadAction<string | null>
    ) => {
      state.currentSimulation.aarligInntektFoerUttakBeloep =
        action.payload !== null ? formatInntekt(action.payload) : null
    },
    setCurrentSimulationAarligInntektVsaHelPensjon: (
      state,
      action: PayloadAction<AarligInntektVsaPensjon | undefined>
    ) => {
      state.currentSimulation.aarligInntektVsaHelPensjon =
        action.payload && action.payload.beloep
          ? {
              ...action.payload,
              beloep: formatInntekt(action.payload?.beloep),
            }
          : undefined
    },
    setCurrentSimulationGradertUttaksperiode: (
      state,
      action: PayloadAction<GradertUttak | null>
    ) => {
      state.currentSimulation.gradertUttaksperiode = action.payload
        ? {
            ...action.payload,
            aarligInntektVsaPensjonBeloep: action.payload
              ? formatInntekt(action.payload.aarligInntektVsaPensjonBeloep)
              : undefined,
          }
        : null
    },
    syncCurrentSimulationFormatertUttaksalderReadOnly: (
      state,
      action: PayloadAction<string | null>
    ) => {
      state.currentSimulation = {
        ...state.currentSimulation,
        formatertUttaksalderReadOnly: action.payload,
      }
    },
    flush: (state) => {
      state.utenlandsopphold = null
      state.samtykke = null
      state.afp = null
      state.samboer = null
      state.currentSimulation = { ...userInputInitialState.currentSimulation }
    },
    flushCurrentSimulation: (state) => {
      state.currentSimulation = { ...userInputInitialState.currentSimulation }
    },
  },
})

export default userInputSlice.reducer

export const userInputActions = userInputSlice.actions

export type UserInputSlice = {
  [userInputSlice.name]: ReturnType<(typeof userInputSlice)['reducer']>
}
