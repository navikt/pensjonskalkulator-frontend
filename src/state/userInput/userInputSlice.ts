import { PayloadAction, createSlice } from '@reduxjs/toolkit'

import { formatInntekt } from '@/utils/inntekt'

export interface Simulation {
  formatertUttaksalderReadOnly: string | null // (!) Obs READONLY - string i format "YY alder.aar string.og M alder.maaneder" - oppdateres automatisk basert på uttaksalder - se uttaksalderListener
  uttaksalder: Alder | null // valgt uttaksalder for 100% alderspensjon (alder perioden gjelder FRA) - aar heltall, maaneder heltall mellom 0-11
  aarligInntektFoerUttakBeloep: string | null // inntekt før uttak av pensjon - formatert string i nok - overskriver beløp fra Skatteetaten
  aarligInntektVsaHelPensjon?: AarligInntektVsaPensjon
  gradertUttaksperiode: GradertUttak | null
  beregningsvalg: Beregningsvalg | null
}

export interface UserInputState {
  veilederBorgerFnr?: string
  veilederBorgerEncryptedFnr?: string
  harUtenlandsopphold: boolean | null
  utenlandsperioder: Utenlandsperiode[]
  samtykke: boolean | null
  samtykkeOffentligAFP: boolean | null
  afp: AfpRadio | null
  skalBeregneAfpKap19: boolean | null
  sivilstand: Sivilstand | null
  epsHarPensjon: boolean | null
  epsHarInntektOver2G: boolean | null
  afpInntektMaanedFoerUttak: boolean | null
  currentSimulation: Simulation
}

export const userInputInitialState: UserInputState = {
  veilederBorgerFnr: undefined,
  veilederBorgerEncryptedFnr: undefined,
  harUtenlandsopphold: null,
  utenlandsperioder: [],
  samtykke: null,
  samtykkeOffentligAFP: null,
  afp: null,
  skalBeregneAfpKap19: null,
  sivilstand: null,
  epsHarInntektOver2G: null,
  epsHarPensjon: null,
  afpInntektMaanedFoerUttak: null,
  currentSimulation: {
    beregningsvalg: null,
    formatertUttaksalderReadOnly: null,
    uttaksalder: null,
    aarligInntektFoerUttakBeloep: null,
    gradertUttaksperiode: null,
  },
}

export const userInputSlice = createSlice({
  name: 'userInputSlice',
  initialState: userInputInitialState,
  reducers: {
    setVeilederBorgerFnr: (
      state,
      action: PayloadAction<{ fnr: string; encryptedFnr: string }>
    ) => {
      state.veilederBorgerFnr = action.payload.fnr
      state.veilederBorgerEncryptedFnr = action.payload.encryptedFnr
    },
    setHarUtenlandsopphold: (state, action: PayloadAction<boolean>) => {
      state.harUtenlandsopphold = action.payload
    },
    setUtenlandsperiode: (state, action: PayloadAction<Utenlandsperiode>) => {
      const index = state.utenlandsperioder.findIndex(
        (item) => item.id === action.payload.id
      )
      if (index !== -1) {
        // Update the existing object
        state.utenlandsperioder[index] = action.payload
      } else {
        // Add the new object
        state.utenlandsperioder.push(action.payload)
      }
    },
    deleteUtenlandsperiode: (state, action: PayloadAction<string>) => {
      state.utenlandsperioder = state.utenlandsperioder.filter(
        (utenlandsperiode) => utenlandsperiode.id !== action.payload
      )
    },
    flushUtenlandsperioder: (state) => {
      state.utenlandsperioder = []
    },
    setSamtykke: (state, action: PayloadAction<boolean>) => {
      state.samtykke = action.payload
    },
    setSamtykkeOffentligAFP: (state, action: PayloadAction<boolean>) => {
      state.samtykkeOffentligAFP = action.payload
    },
    setAfp: (state, action: PayloadAction<AfpRadio>) => {
      state.afp = action.payload
    },
    setSkalBeregneAfpKap19: (state, action: PayloadAction<boolean | null>) => {
      state.skalBeregneAfpKap19 = action.payload
    },
    setAfpInntektMaanedFoerUttak: (
      state,
      action: PayloadAction<boolean | null>
    ) => {
      state.afpInntektMaanedFoerUttak = action.payload
    },
    setSivilstand: (
      state,
      action: PayloadAction<{
        sivilstand: Sivilstand
        epsHarPensjon: boolean | null
        epsHarInntektOver2G: boolean | null
      }>
    ) => {
      state.sivilstand = action.payload.sivilstand
      state.epsHarInntektOver2G = action.payload.epsHarInntektOver2G
      state.epsHarPensjon = action.payload.epsHarPensjon
    },
    setCurrentSimulationBeregningsvalg: (
      state,
      action: PayloadAction<Beregningsvalg | null>
    ) => {
      state.currentSimulation.beregningsvalg = action.payload ?? null
    },
    setCurrentSimulationUttaksalder: (
      state,
      action: PayloadAction<{
        aar: number
        maaneder: number
      } | null>
    ) => {
      state.currentSimulation = {
        ...state.currentSimulation,
        uttaksalder: action.payload ? { ...action.payload } : null,
      }
    },
    setCurrentSimulationAarligInntektFoerUttakBeloep: (
      state,
      action: PayloadAction<string | null>
    ) => {
      state.currentSimulation.aarligInntektFoerUttakBeloep =
        action.payload !== null ? formatInntekt(action.payload) : null
    },
    setCurrentSimulationAarligInntektVsaHelPensjon: (
      state,
      action: PayloadAction<AarligInntektVsaPensjon | undefined>
    ) => {
      state.currentSimulation.aarligInntektVsaHelPensjon = action.payload
        ?.beloep
        ? {
            ...action.payload,
            beloep: formatInntekt(action.payload.beloep),
          }
        : undefined
    },
    setCurrentSimulationGradertUttaksperiode: (
      state,
      action: PayloadAction<GradertUttak | null>
    ) => {
      state.currentSimulation.gradertUttaksperiode = action.payload
        ? {
            ...action.payload,
            aarligInntektVsaPensjonBeloep: action.payload
              ? formatInntekt(action.payload.aarligInntektVsaPensjonBeloep)
              : undefined,
          }
        : null
    },
    syncCurrentSimulationFormatertUttaksalderReadOnly: (
      state,
      action: PayloadAction<string | null>
    ) => {
      state.currentSimulation = {
        ...state.currentSimulation,
        formatertUttaksalderReadOnly: action.payload,
      }
    },
    flush: (state) => {
      state.harUtenlandsopphold = null
      state.utenlandsperioder = []
      state.samtykke = null
      state.samtykkeOffentligAFP = null
      state.afp = null
      state.sivilstand = null
      state.epsHarPensjon = null
      state.epsHarInntektOver2G = null
      state.skalBeregneAfpKap19 = null
      state.afpInntektMaanedFoerUttak = null
      state.currentSimulation = { ...userInputInitialState.currentSimulation }
    },
    flushCurrentSimulation: (state) => {
      state.currentSimulation = userInputInitialState.currentSimulation
    },
  },
})

export default userInputSlice.reducer

export const userInputActions = userInputSlice.actions

export type UserInputSlice = {
  [userInputSlice.name]: ReturnType<(typeof userInputSlice)['reducer']>
}
