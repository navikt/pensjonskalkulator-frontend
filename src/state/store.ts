import {
  combineReducers,
  configureStore,
  createListenerMiddleware,
  ListenerEffectAPI,
  TypedStartListening,
  TypedAddListener,
  PreloadedState,
} from '@reduxjs/toolkit'

import { apiSlice } from './api/apiSlice'
import { createUttaksalderListener } from './listeners/uttaksalderListener'
import userInputReducer from './userInput/userInputReducer'

const listenerMiddleware = createListenerMiddleware()

export const rootReducer = combineReducers({
  [apiSlice.reducerPath]: apiSlice.reducer,
  userInput: userInputReducer,
})

export function setupStore(
  preloadedState?: PreloadedState<RootState>,
  isDev = false
) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware(
        isDev ? { immutableCheck: false, serializableCheck: false } : {}
      )
        .concat(apiSlice.middleware)
        .prepend(listenerMiddleware.middleware),
  })
}

export const store = setupStore()

createUttaksalderListener(
  listenerMiddleware.startListening as AppStartListening
)

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']

export type AppListenerEffectAPI = ListenerEffectAPI<RootState, AppDispatch>
export type AppStartListening = TypedStartListening<RootState, AppDispatch>
export type AppAddListener = TypedAddListener<RootState, AppDispatch>
