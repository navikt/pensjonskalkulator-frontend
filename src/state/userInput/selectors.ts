import { createSelector } from '@reduxjs/toolkit'

import { apiSlice } from '@/state/api/apiSlice'
import { RootState } from '@/state/store'
import { Simulation } from '@/state/userInput/userInputSlice'
import { formatInntekt } from '@/utils/inntekt'
import { isLoependeVedtakEndring } from '@/utils/loependeVedtak'

export const selectHarUtenlandsopphold = (state: RootState): boolean | null =>
  state.userInput.harUtenlandsopphold

export const selectUtenlandsperioder = (state: RootState) =>
  state.userInput.utenlandsperioder

export const selectSamtykke = (state: RootState): boolean | null =>
  state.userInput.samtykke

export const selectSamtykkeOffentligAFP = (state: RootState): boolean | null =>
  state.userInput.samtykkeOffentligAFP

export const selectVeilederBorgerFnr = (state: RootState) =>
  state.userInput.veilederBorgerFnr

export const selectVeilederBorgerEncryptedFnr = (state: RootState) =>
  state.userInput.veilederBorgerEncryptedFnr

export const selectIsVeileder = (state: RootState) =>
  !!state.userInput.veilederBorgerFnr ||
  !!state.userInput.veilederBorgerEncryptedFnr

export const selectAfp = (state: RootState): AfpRadio | null =>
  state.userInput.afp

export const selectFoedselsdato = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    return apiSlice.endpoints.getPerson.select(undefined)(state)?.data
      ?.foedselsdato
  }
)

export const selectNedreAldersgrense = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    return apiSlice.endpoints.getPerson.select(undefined)(state)?.data
      ?.pensjoneringAldre.nedreAldersgrense as Alder
  }
)

export const selectNormertPensjonsalder = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    return apiSlice.endpoints.getPerson.select(undefined)(state)?.data
      ?.pensjoneringAldre.normertPensjoneringsalder as Alder
  }
)

export const selectSivilstand = (state: RootState) => {
  if (state.userInput.sivilstand) {
    return state.userInput.sivilstand
  }

  // Henter sivilstand fra vedtak hvis det er en endringssÃ¸knad, hvis ikke hentes sivilstand fra personopplysninger
  const isEndring = selectIsEndring(state)
  if (isEndring) {
    return apiSlice.endpoints.getLoependeVedtak.select()(state)?.data
      ?.alderspensjon?.sivilstand
  } else {
    const personQuerySivilstandResponse =
      apiSlice.endpoints.getPerson.select()(state).data?.sivilstand

    return personQuerySivilstandResponse
  }
}

export const selectEpsHarInntektOver2G = (state: RootState): boolean | null =>
  state.userInput.epsHarInntektOver2G

export const selectEpsHarPensjon = (state: RootState): boolean | null =>
  state.userInput.epsHarPensjon

export const selectAarligInntektFoerUttakBeloepFraBrukerInput = (
  state: RootState
): string | null =>
  state.userInput.currentSimulation.aarligInntektFoerUttakBeloep

export const selectAarligInntektFoerUttakBeloepFraSkatt = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    const aarligInntektFraSkatt =
      apiSlice.endpoints.getInntekt.select(undefined)(state)?.data
    return aarligInntektFraSkatt
      ? {
          ...aarligInntektFraSkatt,
          beloep: formatInntekt(aarligInntektFraSkatt?.beloep),
        }
      : undefined
  }
)

export const selectAarligInntektFoerUttakBeloep = (
  state: RootState
): string | null | undefined => {
  const aarligInntektFoerUttakBeloepFraBrukerInput =
    selectAarligInntektFoerUttakBeloepFraBrukerInput(state)

  if (aarligInntektFoerUttakBeloepFraBrukerInput === null) {
    return formatInntekt(
      selectAarligInntektFoerUttakBeloepFraSkatt(state, undefined)?.beloep
    )
  }
  return aarligInntektFoerUttakBeloepFraBrukerInput
}

export const selectFormatertUttaksalderReadOnly = (
  state: RootState
): string | null =>
  state.userInput.currentSimulation.formatertUttaksalderReadOnly

export const selectCurrentSimulation = (state: RootState): Simulation =>
  state.userInput.currentSimulation

export const selectHarHentetOffentligTp = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    const offentligTpEntries = Object.keys(state.api.queries).filter((key) =>
      key.startsWith('offentligTp')
    )

    return offentligTpEntries.some(
      (key) => !state.api.queries[key]?.isUninitialized
    )
  }
)

export const selectLoependeVedtak = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    return apiSlice.endpoints.getLoependeVedtak.select(undefined)(state)
      ?.data as LoependeVedtak
  }
)

export const selectUfoeregrad = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    return apiSlice.endpoints.getLoependeVedtak.select(undefined)(state)?.data
      ?.ufoeretrygd.grad as number
  }
)

export const selectIsEndring = createSelector(
  [(state) => state, (_, params = undefined) => params],
  (state) => {
    if (!apiSlice.endpoints.getLoependeVedtak.select(undefined)(state)?.data) {
      return false
    }
    return isLoependeVedtakEndring(
      apiSlice.endpoints.getLoependeVedtak.select(undefined)(state)
        ?.data as LoependeVedtak
    )
  }
)
