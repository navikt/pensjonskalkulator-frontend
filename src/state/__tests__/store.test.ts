import { describe, it } from 'vitest'

import { setupStore } from '../store'
import { UserInputState } from '../userInput/userInputReducer'

describe('store', () => {
  it('returnerer store med riktig slices og default state', () => {
    const store = setupStore()
    const initialState = store.getState()
    expect(initialState).toMatchSnapshot()
  })

  it('returnerer store med riktig state nÃ¥r setupStore kalles med parameter', () => {
    const state: UserInputState = {
      utenlandsopphold: true,
      samtykke: true,
      afp: null,
      samboer: null,
      formatertUttaksalder: null,
      currentSimulation: {
        startAar: null,
        startMaaned: null,
        aarligInntekt: 500000,
        uttaksgrad: 0,
      },
    }

    const store = setupStore({ userInput: state }, true)
    const initialState = store.getState()

    expect(initialState.userInput).toMatchObject(state)
  })
})
