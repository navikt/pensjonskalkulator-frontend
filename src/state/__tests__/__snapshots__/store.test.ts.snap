// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`store > returnerer store med riktig slices og default state 1`] = `
{
  "api": {
    "config": {
      "focused": true,
      "invalidationBehavior": "delayed",
      "keepUnusedDataFor": 3600,
      "middlewareRegistered": false,
      "online": true,
      "reducerPath": "api",
      "refetchOnFocus": false,
      "refetchOnMountOrArgChange": false,
      "refetchOnReconnect": false,
    },
    "mutations": {},
    "provided": {},
    "queries": {},
    "subscriptions": {},
  },
  "userInput": {
    "afp": null,
    "currentSimulation": {
      "aarligInntektFoerUttakBeloep": null,
      "beregningsvalg": null,
      "formatertUttaksalderReadOnly": null,
      "gradertUttaksperiode": null,
      "uttaksalder": null,
    },
    "epsHarInntektOver2G": null,
    "epsHarPensjon": null,
    "harUtenlandsopphold": null,
    "samtykke": null,
    "samtykkeOffentligAFP": null,
    "sivilstand": null,
    "utenlandsperioder": [],
    "veilederBorgerEncryptedFnr": undefined,
    "veilederBorgerFnr": undefined,
  },
}
`;
